packages/solutions/app-tools/src/plugins/deploy/index.ts

```diff
@@ -0,0 +1,68 @@
+import { CliPlugin, IAppContext, NormalizedConfig } from '@modern-js/core';
+import { provider } from 'std-env';
+import { AppTools } from '../../types';
+import { getProjectUsage } from './utils';
+import { createNodePreset } from './platforms/node';
+import { createVercelPreset } from './platforms/vercel';
+import { createNetlifyPreset } from './platforms/netlify';
+
+type DeployPresetCreators = {
+  node: typeof createNodePreset;
+  vercel: typeof createVercelPreset;
+  netlify: typeof createNetlifyPreset;
+};
+
+type DeployTarget = keyof DeployPresetCreators;
+
+const deployPresets: DeployPresetCreators = {
+  node: createNodePreset,
+  vercel: createVercelPreset,
+  netlify: createNetlifyPreset,
+};
+
+async function getDeployPreset(
+  appContext: IAppContext,
+  modernConfig: NormalizedConfig<AppTools>,
+  deployTarget: DeployTarget,
+) {
+  const { appDirectory, distDirectory } = appContext;
+  const { useSSR, useAPI, useWebServer } = getProjectUsage(
+    appDirectory,
+    distDirectory,
+  );
+  const needModernServer = useSSR || useAPI || useWebServer;
+
+  const createPreset = deployPresets[deployTarget];
+
+  if (!createPreset) {
+    throw new Error(
+      `Unknown deploy target: '${deployTarget}'. MODERNJS_DEPLOY should be 'node', 'vercel', or 'netlify'.`,
+    );
+  }
+
+  return createPreset(appContext, modernConfig, needModernServer);
+}
+
+export default (): CliPlugin<AppTools> => ({
+  name: '@modern-js/plugin-deploy',
+  setup: api => {
+    const deployTarget = process.env.MODERNJS_DEPLOY || provider || 'node';
+
+    return {
+      async beforeDeploy() {
+        const appContext = api.useAppContext();
+        const modernConfig = api.useResolvedConfigContext();
+        const deployPreset = await getDeployPreset(
+          appContext,
+          modernConfig,
+          deployTarget as DeployTarget,
+        );
+
+        deployPreset?.prepare && (await deployPreset?.prepare());
```

Comment:
> maybe we should use Object-oriented to develop it hhh

Is it similar to java? I think it's because java isn't flexible enough that we need to write it this way, but we're using JS =、=

Reply:
> And shouldn't we directly write the deploy code in the deploy() function in apptools main process, instead of using plugins and writing it in the hook?

As shown in the code, the hooks needed for deployment are fairly simple, the plugin hooks are more complex, and I didn't want to add some hooks to the plugin core and complicate things in the first version.

Reply:
> As shown in the code, the hooks needed for deployment are fairly simple, the plugin hooks are more complex, and I didn't want to add some hooks to the plugin core and complicate things in the first version.

I have encountered a problem before, which is that the plugin needs to control the order of itself and `appTools`. This often leads to some hooks needing to be executed before `appTools` and some hooks needing to be executed later, So it cannot be implemented. I think this is a problem with our hook and internal logic design.

For the current code, I think `deploy` is our core logic and should not be executed in beforeDeploy..Maybe the better way is:
```
runner.beforeDeploy();
deploy(); // or runner.deploy()
runner.afterDeploy();

```

Reply:
done, add a new `deploy` hook to plugin core